sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java

```diff
@@ -0,0 +1,230 @@
+
+    PAssert.thatSingleton(
+        output.apply("Count All", Count.<KV<String, Integer>>globally()))
+        .isEqualTo(1000L);
+
+    PAssert.that(output
+        .apply("Count Scientist", Count.<String, Integer>perKey())
+    ).satisfies(new SerializableFunction<Iterable<KV<String, Long>>, Void>() {
+      @Override
+      public Void apply(Iterable<KV<String, Long>> input) {
+        for (KV<String, Long> element : input) {
+          assertEquals(100L, element.getValue().longValue());
+        }
+        return null;
+      }
+    });
+
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testReadWithConfiguration() throws Exception {
+    TestPipeline pipeline = TestPipeline.create();
+
+    PCollection<KV<String, Integer>> output = pipeline.apply(
+        JdbcIO.read()
+            .withDriverClassName("org.apache.derby.jdbc.ClientDriver")
+            .withUrl("jdbc:derby://localhost:1527/target/beam")
+            .withQuery("select name,id from BEAM")
+            .withRowMapper(new JdbcIO.RowMapper<KV<String, Integer>>() {
+              @Override
+              public KV<String, Integer> mapRow(ResultSet resultSet) throws Exception {
+                KV<String, Integer> kv = KV.of(resultSet.getString("name"), resultSet.getInt("id"));
+                return kv;
+              }
+            }))
+        .setCoder(KvCoder.of(StringUtf8Coder.of(), SerializableCoder.of(Integer.class)));
+
+    PAssert.thatSingleton(
+        output.apply("Count All", Count.<KV<String, Integer>>globally()))
+        .isEqualTo(1000L);
+
+    PAssert.that(output
+        .apply("Count Scientist", Count.<String, Integer>perKey())
+    ).satisfies(new SerializableFunction<Iterable<KV<String, Long>>, Void>() {
+      @Override
+      public Void apply(Iterable<KV<String, Long>> input) {
+        for (KV<String, Long> element : input) {
+          assertEquals(100L, element.getValue().longValue());
+        }
+        return null;
+      }
+    });
+
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWrite() throws Exception {
+    TestPipeline pipeline = TestPipeline.create();
+
+    ArrayList<KV<Integer, String>> data = new ArrayList<>();
+    for (int i = 0; i < 1000; i++) {
+      KV<Integer, String> kv = KV.of(i, "Test");
+      data.add(kv);
+    }
+    pipeline.apply(Create.of(data))
+        .apply(JdbcIO.write().withDataSource(dataSource)
+            .withStatement("insert into BEAM values(?, ?)")
+            .withPreparedStatementSetter(new JdbcIO.PreparedStatementSetter<KV<Integer, String>>() {
+              public void setParameters(KV<Integer, String> element, PreparedStatement statement)
+                  throws Exception {
+                statement.setInt(1, element.getKey());
+                statement.setString(2, element.getValue());
+              }
+            }));
+
+    pipeline.run();
+
+    Connection connection = dataSource.getConnection();
```

Comment:
These guys need to be closed too. I might have missed more stuff in the test - please make sure you use try-with-resources for all connections, statements and resultsets in this file.